chat/src/main/java/com/github/twitch4j/chat/TwitchChat.java

```diff
@@ -357,6 +388,49 @@ public TwitchChat(EventManager eventManager, CredentialManager credentialManager
                 }
             }
         });
+
+        // Initialize joinAttemptsByChannelName (on an attempt expiring without explicit removal, we retry with exponential backoff)
+        if (maxJoinRetries > 0) {
+            final long initialWait = Math.max(chatQueueTimeout, 5000L);
+            this.joinAttemptsByChannelName = Caffeine.newBuilder()
+                .expireAfterWrite(initialWait, TimeUnit.MILLISECONDS)
+                .scheduler(Scheduler.forScheduledExecutorService(taskExecutor)) // required for prompt removals on java 8
+                .<String, Integer>evictionListener((name, attempts, cause) -> {
+                    if (cause == RemovalCause.EXPIRED && name != null && attempts != null) {
+                        if (attempts < maxJoinRetries) {
+                            taskExecutor.schedule(() -> {
+                                if (currentChannels.contains(name)) {
+                                    issueJoin(name, attempts + 1);
+                                }
+                            }, initialWait * (1L << (attempts + 1)), TimeUnit.MILLISECONDS); // exponential backoff (pow2 optimization)
+                        } else if (removeChannelOnJoinFailure && removeCurrentChannel(name)) {
+                            eventManager.publish(new ChannelRemovedPostJoinFailureEvent(name, ChannelRemovedPostJoinFailureEvent.Reason.RETRIES_EXHAUSTED));
+                        } else {
+                            log.debug("Chat connection exhausted retries when attempting to join channel: {}", name);
+                        }
+                    }
+                })
+                .build();
+        } else {
+            this.joinAttemptsByChannelName = Caffeine.newBuilder().maximumSize(0).build(); // optimization
+        }
+
+        // Remove successfully joined channels from joinAttemptsByChannelName (as further retries are not needed)
+        Consumer<AbstractChannelEvent> joinListener = e -> joinAttemptsByChannelName.invalidate(e.getChannel().getName().toLowerCase());
+        eventManager.onEvent(ChannelStateEvent.class, joinListener::accept);
+        eventManager.onEvent(ChannelNoticeEvent.class, joinListener::accept);
+        eventManager.onEvent(UserStateEvent.class, joinListener::accept);
+
+        // Ban Listener
+        final Set<NoticeTag> banNotices = EnumSet.of(NoticeTag.MSG_BANNED, NoticeTag.MSG_CHANNEL_SUSPENDED, NoticeTag.TOS_BAN); // bit vector
```

Comment:
since there's only three states we care about, is this overengineered?

Reply:
This is a good idea, but we could move this into a static helper class where we could keep similar enum sets for such use cases like first chatter and the user introduction.

Reply:
Sounds good, we can add that class at a later date when there are more enumsets to keep track of